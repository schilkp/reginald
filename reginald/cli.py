from collections.abc import Callable
from dataclasses import dataclass
from typing import List

from typing_extensions import Self

from reginald.datamodel import RegisterMap


@dataclass
class CLI:
    input_file: str
    output_path: str
    generator: Callable[[RegisterMap, Self], None]
    generator_args: List[str]
