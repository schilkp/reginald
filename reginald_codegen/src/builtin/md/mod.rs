use crate::{error::Error, utils::table_col_width};
use std::fmt::Write;

pub mod datasheet;

pub fn md_table(out: &mut dyn Write, rows: &Vec<Vec<String>>) -> Result<(), Error> {
    if rows.len() < 2 {
        return Err(Error::GeneratorError(format!(
            "Cannot generate markdown table from {} rows, need at least 2.",
            rows.len()
        )));
    }

    let col_widths = table_col_width(rows);

    md_table_row(out, &rows[0], &col_widths)?;
    md_sep_row(out, &col_widths)?;
    for col in rows.iter().skip(1) {
        md_table_row(out, col, &col_widths)?;
    }

    Ok(())
}

fn md_table_row(out: &mut dyn Write, row: &[String], widths: &[usize]) -> Result<(), Error> {
    write!(out, "| ")?;
    for (col_idx, width) in widths.iter().enumerate() {
        let col = row.get(col_idx).map_or(String::new(), std::borrow::ToOwned::to_owned);
        if col_idx != 0 {
            write!(out, " | ")?;
        }
        write!(out, "{col: <width$}")?;
    }
    writeln!(out, " |")?;

    Ok(())
}

fn md_sep_row(out: &mut dyn Write, widths: &[usize]) -> Result<(), Error> {
    write!(out, "| ")?;
    for (col_idx, width) in widths.iter().enumerate() {
        if col_idx != 0 {
            write!(out, " | ")?;
        }
        for _ in 0..*width {
            write!(out, "-")?;
        }
    }
    writeln!(out, " |")?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_md_table() {
        let mut is = String::new();

        md_table(
            &mut is,
            &vec![
                vec!["1".into(), "22222".into(), "3".into()],
                vec!["A".into(), "B".into(), "CCCC".into()],
                vec![":)".into(), "!".into()],
            ],
        )
        .unwrap();
        let should = "| 1  | 22222 | 3    |\n\
                      | -- | ----- | ---- |\n\
                      | A  | B     | CCCC |\n\
                      | :) | !     |      |\n";
        assert_eq!(is, should);
    }
}
