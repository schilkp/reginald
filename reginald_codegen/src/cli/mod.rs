mod cmd;
mod diff;

use std::process::ExitCode;

use clap::Parser;

#[derive(Parser, Debug)]
#[command(version, about)]
#[command(name = "reginald")]
enum Cli {
    Gen(cmd::gen::Command),
    Completion(cmd::completion::Command),
    Tool(cmd::tool::Command),
}

pub fn cli_main() -> ExitCode {
    let cli = Cli::parse();

    let err = match cli {
        Cli::Gen(gen) => cmd::gen::cmd(gen),
        Cli::Completion(c) => cmd::completion::cmd(c),
        Cli::Tool(tool) => cmd::tool::cmd(tool),
    };

    match err {
        Ok(_) => ExitCode::SUCCESS,
        Err(err) => {
            eprintln!("{}", err);
            ExitCode::FAILURE
        }
    }
}
