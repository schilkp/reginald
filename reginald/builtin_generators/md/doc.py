import os
from math import inf

from tabulate import tabulate

from reginald.cli import CLI
from reginald.datamodel import *
from reginald.generator import OutputGenerator
from reginald.utils import str_list


class Generator(OutputGenerator):
    @classmethod
    def description(cls):
        return "TODO"

    @classmethod
    def generate(cls, map: RegisterMap, cli: CLI):
        out = []

        # Generate header:
        out.append(f"# {map.map_name} Register Map")
        out.extend(map.docs.multi_line(prefix=""))
        out.append("")

        # Generate overview table:
        out.append(f"## Overview:")
        out.append("")
        rows = []

        def sort_by_adr(x: str) -> float:
            adr = map.registers[x].adr
            if adr is None:
                return inf
            else:
                return adr
        regs = sorted(map.registers.keys(), key=sort_by_adr)

        for reg_name in regs:
            reg = map.registers[reg_name]

            if reg.adr is None:
                adr = "?"
            else:
                adr = f"0x{reg.adr:X}"

            fields = str_list(reg.fields.keys())
            rows.append([adr, reg_name, fields])

        out.append("")
        out.append(tabulate(rows, headers=["Address", "Register", "Fields"], tablefmt="pipe"))
        out.append("")

        # Generate register section:

        out.append(f"## Registers:")
        for reg in map.registers.values():

            # Register name:
            out.append(f"### {reg.name}:")

            # Register info:
            out.extend(reg.docs.two_line(prefix=" -"))
            if reg.adr is not None:
                out.append(f" - Address: 0x{reg.adr:X}")
            if reg.reset_val is not None:
                out.append(f" - Reset Val: 0x{reg.reset_val:X}")

            # Register bitfields table:

            # Collect all bitranges that make up this register - field or not:
            register_bitranges = []  # type: List[BitRange]
            for field in reg.fields.values():
                for range in field.get_bitranges():
                    register_bitranges.append(range)
            register_bitranges.extend(reg.get_unused_bits(include_always_write=True).get_bitranges())

            # Sort bitranges:
            register_bitranges = sorted(register_bitranges, key=lambda x: x.lsb_position, reverse=True)

            bitrow = ["Bits:"]
            field_row = ["Field:"]
            access_row = ["Access:"]

            for bitrange in register_bitranges:
                # Retrieve field that coresponds to this range (if any):
                field_name = reg.get_fieldname_at(bitrange.lsb_position)

                bitrow.append(str(bitrange))

                if field_name is not None:
                    field = reg.fields[field_name]

                    field_row.append(field_name)
                    if field.access is not None:
                        access_row.append(field.access_str())
                    else:
                        access_row.append("?")
                elif reg.is_bit_always_write(bitrange.lsb_position):
                    access_row.append("?")
                    val = reg.get_always_write_value(Bits.from_bitrange(bitrange))
                    field = f"Always write 0x{val:x}"
                    field_row.append(field)
                else:
                    access_row.append("?")
                    field_row.append("?")

            out.append("")
            out.append(tabulate([bitrow, field_row, access_row], headers="firstrow",
                                tablefmt="pipe", numalign="center", stralign="center"))
            out.append("")

            # Field info:
            out.append("")
            out.append(f"*Bitfields*:")

            for field_name, field in reg.fields.items():

                # Access (if any):
                if field.access is not None:
                    access_str = f" [{field.access_str()}]"
                else:
                    access_str = ""

                out.append("")
                out.append(f"  - {field_name}{access_str}:")

                # Documentation (if any):
                out.extend(field.docs.two_line(prefix=" -"))

                # Accepted values (through local or global enum):
                if field.enum is not None:
                    out.append(f"    - Accepts:")
                    for entry in field.enum.entries.values():
                        if entry.docs.brief is not None:
                            out.append(f"      - {entry.name}: 0x{entry.value:X} ({entry.docs.brief})")
                        else:
                            out.append(f"      - {entry.name}: 0x{entry.value:X}")

            # horizontal rule:
            out.append("")
            out.append("---")

        output_file = os.path.join(cli.output_path, f"{map.map_name.lower()}.md")
        with open(output_file, 'w') as outfile:
            outfile.write("\n".join(out))
        print(f"Generated {output_file}...")
