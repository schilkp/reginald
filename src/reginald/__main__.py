import sys

from reginald.cli import parse_args
from reginald.error import ReginaldException
from reginald.input.convert_yaml import YAMLConverter
from reginald.input.parse_yaml import YAML_RegisterMap
from reginald.input.validate_map import MapValidator


def main():

    try:

        # Parse command line args:
        cli, generator = parse_args()

        # Open, parse, and validate input file:
        r = YAML_RegisterMap.from_yaml_file(cli.input_file)
        r = YAMLConverter(r).convert()
        MapValidator(r).validate()

        # Generate output using selected generator:
        generator.generate(r, cli.input_file, cli.output_file, cli.generator_args)

    except ReginaldException as e:
        print(e, file=sys.stderr)
        exit(-1)

    exit(0)


if __name__ == '__main__':
    main()
