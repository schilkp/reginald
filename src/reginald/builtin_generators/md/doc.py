from typing import List

from tabulate import tabulate

from reginald.datamodel import RegisterMap, Bits
from reginald.generator import OutputGenerator
from reginald.utils import str_list


class Generator(OutputGenerator):
    def description(self):
        return "Markdown register documentation."

    def generate(self, rmap: RegisterMap, input_file: str, output_file: str, args: List[str]):
        out = []

        _ = input_file
        _ = args

        registers = []
        for block in rmap.register_blocks.values():
            for template_name, template in block.register_templates.items():
                for instance_name, instance_adr in block.instances.items():
                    register_name = instance_name + template_name
                    register_adr = instance_adr + template.adr

                    registers.append((register_adr, register_name, template))

        registers.sort(key=lambda x: x[0])

        # Generate header:
        out.append(f"# {rmap.map_name} Register Map")
        out.extend(rmap.docs.as_multi_line(prefix=""))
        out.append("")

        # Generate overview table:
        out.append(f"## Overview:")
        out.append("")
        rows = []

        for reg_adr, reg_name, template in registers:

            fields = str_list(template.fields.keys())
            rows.append([hex(reg_adr), reg_name, fields])
        out.append("")
        out.append(tabulate(rows, headers=["Address", "Register", "Fields"], tablefmt="pipe"))
        out.append("")

        # Generate register section:

        out.append(f"## Registers:")
        for reg_adr, reg_name, template in registers:
            # Register name:
            out.append(f"### {reg_name}:")

            # Register info:
            out.extend(template.docs.as_two_line(prefix=" - "))
            out.append(f" - Address: 0x{reg_adr:X}")
            if template.reset_val is not None:
                out.append(f" - Reset Val: 0x{template.reset_val:X}")

            # Register bitfields table:

            # Collect all bitranges that make up this register - field or not:
            register_bitranges = []
            for field in template.fields.values():
                for range in field.get_bitranges():
                    register_bitranges.append(range)
            register_bitranges.extend(template.get_unused_bits(include_always_write=True).get_bitranges())

            # Sort bitranges:
            register_bitranges = sorted(register_bitranges, key=lambda x: x.lsb_position, reverse=True)

            bitrow = ["Bits:"]
            field_row = ["Field:"]
            access_row = ["Access:"]

            for bitrange in register_bitranges:
                # Retrieve field that coresponds to this range (if any):
                field_name = template.get_fieldname_at(bitrange.lsb_position)

                bitrow.append(str(bitrange))

                if field_name is not None:
                    field = template.fields[field_name]

                    field_row.append(field_name)
                    if field.access is not None:
                        access_row.append(field.access_str())
                    else:
                        access_row.append("?")

                elif template.is_bit_always_write(bitrange.lsb_position):
                    access_row.append("")
                    val = template.get_always_write_value(Bits.from_bitrange(bitrange))
                    field = f"Always write 0x{val:x}"
                    field_row.append(field)
                else:
                    access_row.append("?")
                    field_row.append("?")

            out.append("")
            out.append(tabulate([bitrow, field_row, access_row], headers="firstrow",
                                tablefmt="pipe", numalign="center", stralign="center"))
            out.append("")

            # Field info:
            out.append("")
            out.append(f"*Bitfields*:")

            for field_name, field in template.fields.items():

                # Access (if any):
                if len(field.access) > 0:
                    access_str = f" [{field.access_str()}]"
                else:
                    access_str = ""

                out.append("")
                out.append(f"  - {field_name}{access_str}:")

                # Documentation (if any):
                out.extend(field.docs.as_two_line(prefix="    - "))

                # Accepted values (through local or global enum):
                if field.enum is not None:
                    out.append(f"    - Accepts:")
                    for entry in field.enum.entries.values():
                        if entry.docs.brief is not None:
                            out.append(f"      - {entry.name}: 0x{entry.value:X} ({entry.docs.brief})")
                        else:
                            out.append(f"      - {entry.name}: 0x{entry.value:X}")

            # horizontal rule:
            out.append("")
            out.append("---")

        with open(output_file, 'w') as outfile:
            outfile.write("\n".join(out))
