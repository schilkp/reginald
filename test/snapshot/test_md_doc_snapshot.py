import os

import reginald.builtin_generators.md.doc as builtin_md_doc
from reginald.__main__ import generate
from reginald.cli import CLI

SNAPSHOT_DIR = os.path.dirname(__file__)


def test_md_docs_8b(snapshot):
    generator = builtin_md_doc.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_8b.yaml"),
              output_file="md_docs_8b.md",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_md_docs_9b(snapshot):
    generator = builtin_md_doc.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_9b.yaml"),
              output_file="md_docs_9b.md",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_md_docs_32b(snapshot):
    generator = builtin_md_doc.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_32b.yaml"),
              output_file="md_docs_32b.md",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_md_docs_64b(snapshot):
    generator = builtin_md_doc.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_64b.yaml"),
              output_file="md_docs_64b.md",
              generator_args=[],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot
