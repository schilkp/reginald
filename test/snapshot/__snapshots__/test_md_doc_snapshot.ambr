# serializer version: 1
# name: test_md_docs_32b
  '''
  # Snapshotter32B Register Map
  
  ## Overview:
  
  
  | Address   | Register    | Fields                                         |
  |:----------|:------------|:-----------------------------------------------|
  | 0x0       | BASIC       | FIELD0, FIELD1, FIELD2, FIELD3, FIELD4, FIELD5 |
  | 0xf       | COMPLICATED | FIELD1, FIELD2                                 |
  | 0x10      | CH1_CNFG    | Y, A                                           |
  | 0x11      | CH1_STATUS  | STATUS                                         |
  | 0x20      | CH2_CNFG    | Y, A                                           |
  | 0x21      | CH2_STATUS  | STATUS                                         |
  | 0x30      | CH3_CNFG    | Y, A                                           |
  | 0x31      | CH3_STATUS  | STATUS                                         |
  
  ## Registers:
  ### BASIC:
   - Address: 0x0
  
  |  Bits:  |  31-9  |  8  |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:------:|:---:|:------:|:------:|:------:|:------:|:------:|
  | Field:  | FIELD0 |  ?  | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Access: |        |  ?  |        |        |        |        |        |
  
  
  *Bitfields*:
  
    - FIELD0:
  
    - FIELD1:
  
    - FIELD2:
  
    - FIELD3:
  
    - FIELD4:
  
    - FIELD5:
  
  ---
  ### COMPLICATED:
   - Status.
   - More extensive documentation.
   - Address: 0xF
   - Reset Val: 0xFFFFFFFF
  
  |  Bits:  |  31-8  |  7-6   |       5-4        |  3-0   |
  |:-------:|:------:|:------:|:----------------:|:------:|
  | Field:  |   ?    | FIELD1 | Always write 0x1 | FIELD2 |
  | Access: |   ?    |   r    |                  |   r    |
  
  
  *Bitfields*:
  
    - FIELD1 [r]:
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
    - FIELD2 [r]:
      - Another cool field.
      - Important details.
      - Accepts:
        - EN: 0x3 (Is on.)
  
  ---
  ### CH1_CNFG:
   - Address: 0x10
   - Reset Val: 0x0
  
  |  Bits:  |  31-30  |  29-1  |  0  |
  |:-------:|:-------:|:------:|:---:|
  | Field:  |    Y    |   ?    |  A  |
  | Access: |         |   ?    |     |
  
  
  *Bitfields*:
  
    - Y:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH1_STATUS:
   - Address: 0x11
   - Reset Val: 0x0
  
  |  Bits:  |  31-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Access: |   ?    |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH2_CNFG:
   - Address: 0x20
   - Reset Val: 0x0
  
  |  Bits:  |  31-30  |  29-1  |  0  |
  |:-------:|:-------:|:------:|:---:|
  | Field:  |    Y    |   ?    |  A  |
  | Access: |         |   ?    |     |
  
  
  *Bitfields*:
  
    - Y:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH2_STATUS:
   - Address: 0x21
   - Reset Val: 0x0
  
  |  Bits:  |  31-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Access: |   ?    |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH3_CNFG:
   - Address: 0x30
   - Reset Val: 0x0
  
  |  Bits:  |  31-30  |  29-1  |  0  |
  |:-------:|:-------:|:------:|:---:|
  | Field:  |    Y    |   ?    |  A  |
  | Access: |         |   ?    |     |
  
  
  *Bitfields*:
  
    - Y:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH3_STATUS:
   - Address: 0x31
   - Reset Val: 0x0
  
  |  Bits:  |  31-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Access: |   ?    |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  '''
# ---
# name: test_md_docs_64b
  '''
  # Snapshotter64B Register Map
  
  ## Overview:
  
  
  | Address   | Register    | Fields                                         |
  |:----------|:------------|:-----------------------------------------------|
  | 0x0       | BASIC       | FIELD0, FIELD1, FIELD2, FIELD3, FIELD4, FIELD5 |
  | 0xf       | COMPLICATED | FIELD1, FIELD2                                 |
  | 0x10      | CH1_CNFG    | Y, A                                           |
  | 0x11      | CH1_STATUS  | STATUS                                         |
  | 0x20      | CH2_CNFG    | Y, A                                           |
  | 0x21      | CH2_STATUS  | STATUS                                         |
  | 0x30      | CH3_CNFG    | Y, A                                           |
  | 0x31      | CH3_STATUS  | STATUS                                         |
  
  ## Registers:
  ### BASIC:
   - Address: 0x0
  
  |  Bits:  |  63-9  |  8  |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:------:|:---:|:------:|:------:|:------:|:------:|:------:|
  | Field:  | FIELD0 |  ?  | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Access: |        |  ?  |        |        |        |        |        |
  
  
  *Bitfields*:
  
    - FIELD0:
  
    - FIELD1:
  
    - FIELD2:
  
    - FIELD3:
  
    - FIELD4:
  
    - FIELD5:
  
  ---
  ### COMPLICATED:
   - Status.
   - More extensive documentation.
   - Address: 0xF
   - Reset Val: 0xFFFFFFFF
  
  |  Bits:  |  63-8  |  7-6   |       5-4        |  3-0   |
  |:-------:|:------:|:------:|:----------------:|:------:|
  | Field:  |   ?    | FIELD1 | Always write 0x1 | FIELD2 |
  | Access: |   ?    |   r    |                  |   r    |
  
  
  *Bitfields*:
  
    - FIELD1 [r]:
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
    - FIELD2 [r]:
      - Another cool field.
      - Important details.
      - Accepts:
        - EN: 0x3 (Is on.)
  
  ---
  ### CH1_CNFG:
   - Address: 0x10
   - Reset Val: 0x0
  
  |  Bits:  |  63-30  |  29-1  |  0  |
  |:-------:|:-------:|:------:|:---:|
  | Field:  |    Y    |   ?    |  A  |
  | Access: |         |   ?    |     |
  
  
  *Bitfields*:
  
    - Y:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH1_STATUS:
   - Address: 0x11
   - Reset Val: 0x0
  
  |  Bits:  |  63-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Access: |   ?    |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH2_CNFG:
   - Address: 0x20
   - Reset Val: 0x0
  
  |  Bits:  |  63-30  |  29-1  |  0  |
  |:-------:|:-------:|:------:|:---:|
  | Field:  |    Y    |   ?    |  A  |
  | Access: |         |   ?    |     |
  
  
  *Bitfields*:
  
    - Y:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH2_STATUS:
   - Address: 0x21
   - Reset Val: 0x0
  
  |  Bits:  |  63-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Access: |   ?    |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH3_CNFG:
   - Address: 0x30
   - Reset Val: 0x0
  
  |  Bits:  |  63-30  |  29-1  |  0  |
  |:-------:|:-------:|:------:|:---:|
  | Field:  |    Y    |   ?    |  A  |
  | Access: |         |   ?    |     |
  
  
  *Bitfields*:
  
    - Y:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH3_STATUS:
   - Address: 0x31
   - Reset Val: 0x0
  
  |  Bits:  |  63-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Access: |   ?    |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  '''
# ---
# name: test_md_docs_8b
  '''
  # Snapshotter8B Register Map
  
  ## Overview:
  
  
  | Address   | Register    | Fields                                 |
  |:----------|:------------|:---------------------------------------|
  | 0x0       | BASIC       | FIELD1, FIELD2, FIELD3, FIELD4, FIELD5 |
  | 0xf       | COMPLICATED | FIELD1, FIELD2                         |
  | 0x10      | CH1_CNFG    | A                                      |
  | 0x11      | CH1_STATUS  | STATUS                                 |
  | 0x20      | CH2_CNFG    | A                                      |
  | 0x21      | CH2_STATUS  | STATUS                                 |
  | 0x30      | CH3_CNFG    | A                                      |
  | 0x31      | CH3_STATUS  | STATUS                                 |
  
  ## Registers:
  ### BASIC:
   - Address: 0x0
  
  |  Bits:  |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:------:|:------:|:------:|:------:|:------:|
  | Field:  | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Access: |        |        |        |        |        |
  
  
  *Bitfields*:
  
    - FIELD1:
  
    - FIELD2:
  
    - FIELD3:
  
    - FIELD4:
  
    - FIELD5:
  
  ---
  ### COMPLICATED:
   - Status.
   - More extensive documentation.
   - Address: 0xF
   - Reset Val: 0x0
  
  |  Bits:  |  7-6   |       5-4        |  3-0   |
  |:-------:|:------:|:----------------:|:------:|
  | Field:  | FIELD1 | Always write 0x1 | FIELD2 |
  | Access: |   r    |                  |   r    |
  
  
  *Bitfields*:
  
    - FIELD1 [r]:
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
    - FIELD2 [r]:
      - Another cool field.
      - Important details.
      - Accepts:
        - EN: 0x3 (Is on.)
  
  ---
  ### CH1_CNFG:
   - Address: 0x10
   - Reset Val: 0x0
  
  |  Bits:  |  7-1  |  0  |
  |:-------:|:-----:|:---:|
  | Field:  |   ?   |  A  |
  | Access: |   ?   |     |
  
  
  *Bitfields*:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH1_STATUS:
   - Address: 0x11
   - Reset Val: 0x0
  
  |  Bits:  |  7-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Access: |   ?   |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH2_CNFG:
   - Address: 0x20
   - Reset Val: 0x0
  
  |  Bits:  |  7-1  |  0  |
  |:-------:|:-----:|:---:|
  | Field:  |   ?   |  A  |
  | Access: |   ?   |     |
  
  
  *Bitfields*:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH2_STATUS:
   - Address: 0x21
   - Reset Val: 0x0
  
  |  Bits:  |  7-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Access: |   ?   |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH3_CNFG:
   - Address: 0x30
   - Reset Val: 0x0
  
  |  Bits:  |  7-1  |  0  |
  |:-------:|:-----:|:---:|
  | Field:  |   ?   |  A  |
  | Access: |   ?   |     |
  
  
  *Bitfields*:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH3_STATUS:
   - Address: 0x31
   - Reset Val: 0x0
  
  |  Bits:  |  7-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Access: |   ?   |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  '''
# ---
# name: test_md_docs_9b
  '''
  # Snapshotter9B Register Map
  
  ## Overview:
  
  
  | Address   | Register    | Fields                                         |
  |:----------|:------------|:-----------------------------------------------|
  | 0x0       | BASIC       | FIELD0, FIELD1, FIELD2, FIELD3, FIELD4, FIELD5 |
  | 0xf       | COMPLICATED | FIELD1, FIELD2                                 |
  | 0x10      | CH1_CNFG    | A                                              |
  | 0x11      | CH1_STATUS  | STATUS                                         |
  | 0x20      | CH2_CNFG    | A                                              |
  | 0x21      | CH2_STATUS  | STATUS                                         |
  | 0x30      | CH3_CNFG    | A                                              |
  | 0x31      | CH3_STATUS  | STATUS                                         |
  
  ## Registers:
  ### BASIC:
   - Address: 0x0
  
  |  Bits:  |   8    |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:------:|:------:|:------:|:------:|:------:|:------:|
  | Field:  | FIELD0 | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Access: |        |        |        |        |        |        |
  
  
  *Bitfields*:
  
    - FIELD0:
  
    - FIELD1:
  
    - FIELD2:
  
    - FIELD3:
  
    - FIELD4:
  
    - FIELD5:
  
  ---
  ### COMPLICATED:
   - Status.
   - More extensive documentation.
   - Address: 0xF
   - Reset Val: 0x0
  
  |  Bits:  |  8-6   |       5-4        |  3-0   |
  |:-------:|:------:|:----------------:|:------:|
  | Field:  | FIELD1 | Always write 0x1 | FIELD2 |
  | Access: |   r    |                  |   r    |
  
  
  *Bitfields*:
  
    - FIELD1 [r]:
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
    - FIELD2 [r]:
      - Another cool field.
      - Important details.
      - Accepts:
        - EN: 0x3 (Is on.)
  
  ---
  ### CH1_CNFG:
   - Address: 0x10
   - Reset Val: 0x0
  
  |  Bits:  |  8-7  |  6-0  |
  |:-------:|:-----:|:-----:|
  | Field:  |   A   |   ?   |
  | Access: |       |   ?   |
  
  
  *Bitfields*:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH1_STATUS:
   - Address: 0x11
   - Reset Val: 0x0
  
  |  Bits:  |  8-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Access: |   ?   |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH2_CNFG:
   - Address: 0x20
   - Reset Val: 0x0
  
  |  Bits:  |  8-7  |  6-0  |
  |:-------:|:-----:|:-----:|
  | Field:  |   A   |   ?   |
  | Access: |       |   ?   |
  
  
  *Bitfields*:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH2_STATUS:
   - Address: 0x21
   - Reset Val: 0x0
  
  |  Bits:  |  8-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Access: |   ?   |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  ### CH3_CNFG:
   - Address: 0x30
   - Reset Val: 0x0
  
  |  Bits:  |  8-7  |  6-0  |
  |:-------:|:-----:|:-----:|
  | Field:  |   A   |   ?   |
  | Access: |       |   ?   |
  
  
  *Bitfields*:
  
    - A:
      - Controls value A
      - Accepts:
        - ON: 0x1
        - OFF: 0x0
  
  ---
  ### CH3_STATUS:
   - Address: 0x31
   - Reset Val: 0x0
  
  |  Bits:  |  8-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Access: |   ?   |        |
  
  
  *Bitfields*:
  
    - STATUS:
      - Status
      - Accepts:
        - COOL: 0x1 (All cool.)
        - NOT_COOL: 0x2 (All not so cool.)
        - HOT: 0x3 (All pretty hot.)
  
  ---
  '''
# ---
